### PR TITLE
adding kv_cache quantization

### PR DESCRIPTION
**Pull request by HDCharles**
Adding support for kv_cache quantization, we are using simple symmetric quantization, though using the full precision of the k and v values of the current token. 

we see tok/s reduction of 3-5 tok/s depending on context length.

image

and a reduction in peak memory

image

We expect this reduction to scale to large context lengths, in the model memory trace we can see the point where we replace the bf16 cache with the int8 cache which visually saves about half of the used memory

Screenshot 2024-08-02 at 2 45 14 AM

at longer context lengths both quantized and non-quantized kv_cache models start outputing weird stuff but otherwise accuracy of the kv_cache quant looks reasonable though e.g. for 2048 context length:

<|begin_of_text|>Hello, my name is Richard Brown and I have been a professional musician for over 25 years. I have played in a number of bands, doing a wide variety of genres (soul/funk, rock, jazz, blues, latin, world). I have played on over a hundred albums so far.
I have played with many different singers, as well as instrumentalists (guitarists, sax players, brass players, etc.). I love to play and try to learn as much as I can from others. I have become an all-round musician - playing keyboards, drums, programming, arranging; as well as writing songs myself. I have my own studio, and I can do sessions online.
I also have my own website, where you can find out more about me and my music.
I hope that you will find the music that you are looking for here.

Otherwise there are some fixes in generate.py to get things working for large context lengths without overflowing beyond the model limit.

test plan: 

sh benchmarks.sh 

(specifically the last 6 rows of benchmark_results.txt)

### CODE REVIEW

**Comment by msaroufim** [comment]
Memory savings are small because the default sequence length is still quite small, I'd increase it substantially to 100K, 200K, 500K, 1M and see how results change then

**Comment by HDCharles** [comment]
> Were the memory decreases also minimal at larger sequence lengths? I'd increase it substantially to 100K, 200K, 500K, 1M and see how results change then

i'll test that

File: torchao/_models/llama/model.py [hunk 1]
```diff
@@ -97,6 +96,53 @@ def update(self, input_pos, k_val, v_val):
...
+        self.k_cache_scale[:, :, input_pos] = k_scale.unsqueeze(-1)
+        del k_val
+
+        q_v_val, v_scale = quantize_activation_per_token_absmax(v_val)
+        self.k_cache[:, :, input_pos] = q_v_val
+        self.k_cache_scale[:, :, input_pos] = v_scale.unsqueeze(-1)
+        del v_val
+
+        # return k_out, v_out
+        return self.k_cache*self.k_cache_scale, self.v_cache*self.v_cache_scale
```
Comment (gau-nernst) [inline torchao/_models/llama/model.py:135]: I'm guessing the bfloat16 kv cache is materialized in global memory, thus no memory saving. need to inspect it further.

**Review by msaroufim — APPROVED** [review]
discussed followup work offline

**Pull request merged**